### PR TITLE
Run build

### PR DESCRIPTION
**Pull request by davidbuzinski**
Initial "Run MATLAB Build" integration for buildtool
**For Freestyle project:**
- Run MATLAB Build option with Tasks parameter (String) that takes a space-separated list of buildtool tasks

**For DSL:**
- runMATLABBuild step with tasks parameter (String) that takes a space-separated list of buildtool tasks

### CODE REVIEW

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import hudson.EnvVars;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.Launcher.ProcStarter;
+import hudson.model.Result;
+import hudson.model.TaskListener;
+
+public class MatlabBuildStepExecution extends SynchronousNonBlockingStepExecution<Void> implements MatlabBuild {
+    
+    private static final long serialVersionUID = 6704588180717665100L;
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:20]: Not sure where to get this from. It might need updated. @nikhilbhoski

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        super(context);
+        this.tasks = tasks;
+    }
+
+    private String getTasks() {
+        return this.tasks;
+    }
+
+    @Override
+    public Void run() throws Exception {
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:34]: This is almost the same in all of our Matlab(Action)StepExecutions. Might be a way to reuse or abstract in the future

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        getContext().setResult(Result.SUCCESS);
+        return null;
+    }
+
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        getContext().onFailure(cause);
+    }
+    
+    private int execMatlabCommand(FilePath workspace, Launcher launcher,
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:60]: This is almost identical in several files. Might be a way to abstract or reuse in the future

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    
+    private void createMatlabScriptByName(FilePath uniqeTmpFolderPath, String uniqueScriptName, FilePath workspace, TaskListener listener) throws IOException, InterruptedException {
+
+        // Create a new command runner script in the temp folder.
+        final FilePath matlabBuildFile =
+                new FilePath(uniqeTmpFolderPath, uniqueScriptName + ".m");
+        final String tasks = getContext().get(EnvVars.class).expand(getTasks());
+        String cmd = "buildtool";
+
+        if (!tasks.trim().isEmpty()) {
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:95]: I don't love that the logic to make the MATLAB command is mixed with the logic to create script in the `createMatlabScriptByName` method

File: src/main/java/com/mathworks/ci/RunMatlabBuildBuilder.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+
+        buildResult = execMatlabCommand(workspace, launcher, listener, env);
+
+        if (buildResult != 0) {
+            build.setResult(Result.FAILURE);
+        }
+    }
+
+    private int execMatlabCommand(FilePath workspace, Launcher launcher,
```
Comment (davidbuzinski) [inline src/main/java/com/mathworks/ci/RunMatlabBuildBuilder.java:93]: Lots of duplication in the `execMatlabCommand` and `createMatlabScriptByName` methods in all 6 of the Builder and StepExecution files. Something to consider going forward

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+        scriptBuilder.setTasks("$TASKS");
+        project.getBuildersList().add(scriptBuilder);
+        FreeStyleBuild build = project.scheduleBuild2(0).get();
+        jenkins.assertLogContains("buildtool compile", build);
+    }
+    
+    /*
+     * Test to verify if appropriate MATLAB runner file is copied in workspace.
+     * 
+     * NOTE: This test assumes there is no MATLAB installed and is not on System Path.
```
Comment (davidbuzinski) [inline src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java:243]: Would love to remove this requirement in the future (in all other tests as well). It took me a while to figure out why tests were failing when I first tried to build.

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+        scriptBuilder.setTasks("");
+        project.getBuildersList().add(scriptBuilder);
+        FreeStyleBuild build = project.scheduleBuild2(0).get();
+        jenkins.assertLogContains("MatlabNotFoundError", build);
+    }
+    
+	/*
+	 * Test to verify if Matrix build fails when MATLAB is not available.
+     * 
+     * NOTE: This test assumes there is no MATLAB installed and is not on System Path.
```
Comment (davidbuzinski) [inline src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java:272]: Same here. I'd love to remove this requirement

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTester.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+import hudson.Launcher.ProcStarter;
+import hudson.model.AbstractProject;
+import hudson.model.Result;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import net.sf.json.JSONObject;
+
+public class RunMatlabBuildBuilderTester extends RunMatlabBuildBuilder {
```
Comment (davidbuzinski) [inline src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTester.java:26]: There is a Tester for each of the types of MATLAB tasks that are almost all the same. We might be able to consolidate here going forward as well (Mock MATLAB executor for all of them?)

File: src/test/java/com/mathworks/ci/RunMatlabBuildStepTest.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        j.assertBuildStatusSuccess(build);
+    }
+
+    /*
+     * Verify appropriate task is invoked as in pipeline script
+     */
+    @Test
+    public void verifyTasksSameAsScript() throws Exception {
+        project.setDefinition(
+                new CpsFlowDefinition("node { runMATLABBuild(tasks: 'compile') }", true));
```
Comment (davidbuzinski) [inline src/test/java/com/mathworks/ci/RunMatlabBuildStepTest.java:96]: A lot of these tests (and the builder tests) feel more like integration tests, and several of them take a while to run. Maybe some abstractions can make the code more directly testable

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java (hunk 1 above)
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:20]: Most IDE's provide auto generation of this UID like eclipse and this should be unique as during serelizations classes are identified with these ID's for now you could safly use `1L` since other values are been auto generated

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java (hunk 6 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java:243]: the rationale behind  tests is  considering the environment of jenkins.  where it would run without actual MATLAB instances. Hence whatever we could test which is closer to actual would be good idea

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java (hunk 7 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java:272]: Same as above

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTester.java (hunk 8 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTester.java:26]: +1

File: src/main/java/com/mathworks/ci/RunMatlabBuildBuilder.java (hunk 5 above)
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/RunMatlabBuildBuilder.java:93]: +1

File: src/main/resources/com/mathworks/ci/RunMatlabBuildBuilder/help-tasks.html [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<div>
+    Insert a space-separated list of tasks to run in the <b>Tasks</b> box. If not specified, the action runs the default tasks in buildfile.m as well as all the tasks on which they depend.<br>
+    <b>Example: </b>test<br>
+    <b>Example: </b>compile test
+    <br>&nbsp;</br>
+    <b>Note:</b>
+    <ul>
+        <li>MATLAB exits with exit code 0 if the tasks run without error. Otherwise, MATLAB terminates with a nonzero exit code, which causes the workflow to fail.</li>
```
Comment (nbhoski) [inline src/main/resources/com/mathworks/ci/RunMatlabBuildBuilder/help-tasks.html:8]: Does it make sense to replace workflow with build.

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java (hunk 6 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java:243]: I also think we could restrict the test cases related to RunMatlabBuilderBuild Class here instead of re running duplicates.

File: src/test/java/com/mathworks/ci/RunMatlabBuildStepTest.java (hunk 9 above)
Comment (nbhoski) [reply src/test/java/com/mathworks/ci/RunMatlabBuildStepTest.java:96]: alternately as i commented earlier we could just restrict test related to RunMatlabBuildstep class.

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java (hunk 3 above)
Comment (nbhoski) [reply src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:60]: Yeah we could declare in MatlabBuild interface

File: src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java (hunk 6 above)
Comment (davidbuzinski) [reply src/test/java/com/mathworks/ci/RunMatlabBuildBuilderTest.java:243]: I am having trouble understanding which duplicate tests to remove. `verifyMATLABrunnerFileGenerated` and `verifyMatrixBuildFails`?

File: src/main/resources/com/mathworks/ci/RunMatlabBuildBuilder/help-tasks.html (hunk 10 above)
Comment (davidbuzinski) [reply src/main/resources/com/mathworks/ci/RunMatlabBuildBuilder/help-tasks.html:8]: Yes good catch

File: src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java (hunk 1 above)
Comment (davidbuzinski) [reply src/main/java/com/mathworks/ci/MatlabBuildStepExecution.java:20]: Ok, I was able to install Eclipse and generate one. For some reason VS Code wasn't generating one for me, even with a Java language extension.
